### PR TITLE
Search/create search logic

### PR DESCRIPTION
**Pull request by PaulTimothyChambers**
**What (if any) features are you implementing?**
- Search functionality

**What (if anything) did you refactor?**
- The index.html file to include script.js in the script tag

**Were there any issues that arose?**
- N/A

**Any other comments, questions, or concerns?**
- N/A

**Is there anything you need from your team?**
- N/A

### CODE REVIEW

File: src/classes/User.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    this.name = usersData.name;
+    this.id = usersData.id;
+    this.pantry = usersData.pantry
+    this.favoriteRecipes = [];
+    this.weeklyFavorites = [];
+    this.filteredByName = [];
+  }
+
+  addFavorite(recipe) {
+    let favRecipe = this.favoriteRecipes;
```
Comment (HartiganHM) [inline src/classes/User.js:14]: All of these `let` variables can be `const`. Using `Array.push` doesn't re-assing the variable, it just updates it, so `const` makes the most sense here.

File: src/classes/User.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  //   recipes.forEach((recipe, i) => {
+  //     let ingredients[i] = recipe.getIngredientsInformation(ingredientsData);
+  //     console.log(ingredients[i])
+  //     ingredients.forEach((ingredient, i) => {
+  //       if (ingredient.id === ingredientsData[i].id) {
+  //         ingredient.name = ingredientsData[i].name;
+  //       }
+  //     })
+  //   })
+  // }
```
Comment (HartiganHM) [inline src/classes/User.js:61]: In general, if code isn't being used, you should remove it. I usually like to have a `scratchpad.js` file locally that's added to my `.gitignore` which lets me hold on to work-in-progress code and test things out without pushing it up to the actual codebase.

File: src/classes/User.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+          acc.push(recipe);
+        }
+      });
+      return acc;
+    }, []);
+    return recipes;
+  }
+
+  filterFavoriteRecipesByIngredient(ingredientName, ingredientsData) {
+    this.getIngredientNames( ingredientsData);
```
Comment (HartiganHM) [inline src/classes/User.js:46]: Looks like you have an extra space before `ingredientsData`.

File: src/data/ingredients-sample.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
+export const ingredientsData = [
```
Comment (HartiganHM) [inline src/data/ingredients-sample.js:1]: It looks like these files are in your other PR, too. Depending on which one gets merged first, just make sure you update the one that hasn't been merged.

File: src/index.html [hunk 5]
```diff
@@ -318,4 +318,5 @@ <h2>INSTRUCTIONS</h2>
       </section>
     </main>
   </body>
+  <script src="./scripts.js"></script>
```
Comment (HartiganHM) [inline src/index.html:321]: I don't think this is needed when using webpack. Try removing it and see if the application still runs as expected with `npm run start`. I could totally be wrong though 🙂

File: src/scripts.js [hunk 6]
```diff
@@ -3,6 +3,9 @@ import apiCalls from './apiCalls';
 import { recipeData } from './data/recipes';
 import RecipeRepository from './classes/RecipeRepository';
 
+// let recipesList = recipesData;  need to plug in recipeData here
```
Comment (HartiganHM) [inline src/scripts.js:6]: Again, comments should be removed for PRs unless they're providing specific in-code context that would help other devs.

File: src/scripts.js [hunk 7]
```diff
@@ -16,9 +19,27 @@ const favoriteRecipesSection = document.getElementById('favoriteRecipesSection')
 
 const weeklyRecipesSection = document.getElementById('weeklyRecipesSection');
 
+let searchBar = document.querySelector('.favorites-view__favorites');
+let mainGlide = document.querySelector('.search-ingredient-glide__images-container');
```
Comment (HartiganHM) [inline src/scripts.js:23]: These should also be `const` unless they're reassigned somewhere. From what I can see, it doesn't look like they are.

File: src/scripts.js [hunk 8]
```diff
@@ -16,9 +19,27 @@ const favoriteRecipesSection = document.getElementById('favoriteRecipesSection')
...
 
 homeViewButton.addEventListener('click', displayHomeView);
 allRecipesButton.addEventListener('click', displayAllRecipes);
+searchBar.addEventListener('keyup', (e) => {
```
Comment (HartiganHM) [inline src/scripts.js:27]: Everything here should be broken out into a separate function versus inline with the event listener.

Also, we have a rule on our team that variable names should be at least three characters long, so we never use `e` and just spell out `event`. This is more preference than anything, but, as you start to work in the field, it's helpful to have _more_ context than less.

File: src/scripts.js [hunk 9]
```diff
@@ -16,9 +19,27 @@ const favoriteRecipesSection = document.getElementById('favoriteRecipesSection')
...
+  mainGlide.innerHTML = '';
+  const searchTerm = e.target.value;
```
Comment (HartiganHM) [inline src/scripts.js:29]: I would update this to use `.toLowerCase()` with the variable definitions so you don't have to use it three times below (e.g. `const searchTerm = e.target.value.toLowerCase();`).

File: src/scripts.js [hunk 10]
```diff
@@ -16,9 +19,27 @@ const favoriteRecipesSection = document.getElementById('favoriteRecipesSection')
...
+  filteredRecipes = recipesList.filter(recipe => {
+    return recipe.name.toLowerCase().includes(searchTerm.toLowerCase());
+    || recipe.tags.toString().toLowerCase().includes(searchTerm.toLowerCase());
+    || recipe.ingredients.toString().toLowerCase().includes(searchTerm.toLowerCase());
```
Comment (HartiganHM) [inline src/scripts.js:33]: There's _a lot_ going on in this return statement that makes it hard to read. I would suggest setting all of these to variables and then returning something easier to read. Something like:

```javascript
const recipeIncludesName = recipe.name.toLowerCase().includes(searchTerm);
...

return recipeIncludesName || ... ;
```

File: src/scripts.js [hunk 11]
```diff
@@ -16,9 +19,27 @@ const favoriteRecipesSection = document.getElementById('favoriteRecipesSection')
...
+  });
+  filteredRecipes.forEach((recipe) => {
```
Comment (HartiganHM) [inline src/scripts.js:35]: This could also be broken out into a dedicated function.

File: test/User-test.js [hunk 12]
```diff
@@ -1,31 +1,31 @@
...
+import { ingredientsData } from '../src/data/ingredients-sample';
 
 describe('User', function() {
-  let user;
-  let recipeOne;
-  let recipeTwo;
-  let recipeThree;
-  let tagOne;
-  let tagTwo;
+  let user, recipeOne, recipeTwo, recipeThree, tagOne, tagTwo, tagThree, ingredientsOne, ingredientsTwo, ingredientsThree, nameOne;
```
Comment (HartiganHM) [inline test/User-test.js:9]: You could take this as a preference, but I wouldn't advise updating to this type of syntax. I personally find it more confusing and harder to read, plus, even though it's a "one-liner", it's a _very long_ one-liner. I've also never seen anyone use this on the job, for what it's worth.

File: test/User-test.js [hunk 13]
```diff
@@ -1,31 +1,31 @@
...
 
-  beforeEach(() => {
-    user = new User(users[0]);
-    recipeOne = new Recipe(recipeData[7]);
-    recipeTwo = new Recipe(recipeData[3]);
+  beforeEach(function() {
```
Comment (HartiganHM) [inline test/User-test.js:11]: No need for this change, the arrow function should work fine.

File: test/User-test.js [hunk 14]
```diff
@@ -1,31 +1,31 @@
...
+    user = new User(usersData[0]);
+    recipeOne = new Recipe(recipeData[2]);
+    recipeTwo = new Recipe(recipeData[1]);
     recipeThree = new Recipe(recipeData[0]);
     tagOne = 'appetizer'
     tagTwo = 'snack'
+    tagThree = 'side dish'
+    nameOne = 'cookies'
```
Comment (HartiganHM) [inline test/User-test.js:19]: Just for clarity on how `beforeEach` works, it allows you to "reset" things before each test. This is useful for things that hold onto "state" like classes, which is why `beforeEach` is awesome for stuff like `user, recipeOne, recipeTwo, etc.`. However, `tagOne, tageTwo, nameOne, etc.` never change. So, you couldn't technically make these `const` outside of the `beforeEach` or even outside of the `describe` block.

No real need to change, but food for thought.

File: test/User-test.js [hunk 15]
```diff
@@ -1,31 +1,31 @@
...
+    // ingredientsOne = recipeOne.getIngredientsInformation()
+    // ingredientsTwo = recipeTwo.getIngredientsInformation()
+    // ingredientsThree = recipeThree.getIngredientsInformation()
```
Comment (HartiganHM) [inline test/User-test.js:22]: Again, if it isn't being used, let's remove it 🙂

File: test/User-test.js [hunk 16]
```diff
@@ -52,29 +52,25 @@ describe('User', function() {
...
-    expect(user.filteredByIngredient).to.deep.equal([recipeOne, recipeThree]);
+  it.skip('should be able to filter favorites by ingredient', function() {
+    expect(user.filterFavoriteRecipesByIngredient(['wheat flour'], ingredientsData)).to.deep.equal([recipeOne, recipeThree]);
   });
 
-  it('should be able to filter favorites by name', function() {
-    // the test portion commented out is the less spicy version; the test not commented out is SUPER spicy, ngl
-    user.filterFavoriteRecipesByName('cookies'); //or => ('Loaded Chocolate Chip Pudding Cookie Cups');
-    expect(user.filteredByName).to.deep.equal([recipeOne, recipeThree]); //or => ([recipeOne]);
+  it.skip('should be able to filter favorites by name', function() {
```
Comment (HartiganHM) [inline test/User-test.js:71]: Did you mean to leave these `skip` statements here?

**Review by HartiganHM — CHANGES_REQUESTED** [review]
Overall, this looks really good! Nice use of `reduce` and `filter`! I left some notes about things that can/should be changed. At a high level, I think there are some refactors you made that don't necessarily make things easier to follow, even if they technically reduce the number of lines of code.

Also, this PR is titled **Search/create search logic**, but there are updates to tests. In general, it's good to keep your PR scoped to what it's about, so to me, the tests would be considered "scope creep" (aka doing extra work outside of what's needed). This isn't necessarily a problem, but, if you do this, make sure to call it out in your PR description OR open it in a separate PR 🙂

**Comment by PaulTimothyChambers** [comment]
My mess

**Pull request closed without merging**